src/main/java/Tree/RedBlackTree.java

```diff
@@ -0,0 +1,372 @@
+            } else if (parent.getRightNode() != null) {
+                sibling = (RedBlackNode<T>) parent.getRightNode();
+            }
+        }
+        return sibling;
+    }
+
+    public void insert(int insert_key, T value) {
+        if (root == null) {
+            RedBlackNode<T> node = new RedBlackNode<T>(insert_key, value, true);
+            current = node;
+            root = (RedBlackNode<T>) current;
+            return;
+        }
+
+        RedBlackNode<T> insertedNode = null;
+        current = root;
+
+        while (true) {
+            if (current.getKey() == insert_key) {
+                current.setValue(value);
+                break;
+            }
+            if (current.getKey() < insert_key) {
+                if (current.getRightNode() == null) {
+                    RedBlackNode<T> node = new RedBlackNode<T>(insert_key, value, true);
+                    linkNodes(current, node, 'R');
+                    insertedNode = node;
+                    break;
+                } else {
+                    moveToRightNode();
+                }
+            } else if (current.getKey() > insert_key) {
+                if (current.getLeftNode() == null) {
+                    RedBlackNode<T> node = new RedBlackNode<T>(insert_key, value, true);
+                    linkNodes(current, node, 'L');
+                    insertedNode = node;
+                    break;
+                } else {
+                    moveToLeftNode();
+                }
+            }
+        }
+        adjustAfterInsert(insertedNode);
+        current = root;
+    }
+
+    public void adjustAfterInsert(RedBlackNode<T> node) {
+        assert node == null : "Input node must not be null";
+
+        // colour the node red
+        node.setColour(false);
+
+        RedBlackNode<T> gp = getGrandParent(node);
+        RedBlackNode<T> parent = (RedBlackNode<T>) node.getParentNode();
+        RedBlackNode<T> uncle = getUncle(node);
+
+        if (parent != null && !parent.getColour()) {
+
+            if (uncle != null && !uncle.getColour()) {
+                parent.setColour(true);
+                uncle.setColour(true);
+                gp.setColour(false);
+                adjustAfterInsert(gp);
+            }
+            else if (parent.getKey() < gp.getKey()) {
+                if (node.getKey() > parent.getKey()) {
+                    leftRotation(parent);
+                }
+                parent.setColour(true);
+                gp.setColour(false);
+                rightRotation(gp);
+            }
+            else if (parent.getKey() >  gp.getKey()) {
+                if (node.getKey() < parent.getKey()) {
+                    rightRotation(parent);
+                }
+                parent.setColour(true);
+                gp.setColour(false);
+                leftRotation(gp);
+            }
+        }
+        RedBlackNode<T> rt = (RedBlackNode<T>) root;
+        rt.setColour(true);
+        root = rt;
```

Comment:
why not just root.setColor(true)?

Also, what if root changes due to rotation??

Reply:
left/right rotation handle root changing cases

Reply:
cannot cast setColour(RedBlackNode function) on BST node root.